cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -139,6 +142,33 @@ indexBaseName repo = repoLocalDir repo </> fn
 -- Reading the source package index
 --
 
+-- 'IndexState' defined in "Distribution.Client.IndexUtils.Timestamp"
+-- to avoid import cycles
+
+-- | Effective 'IndexState' after applying 'filerCache'
+data IndexStateInfo = IndexStateInfo
+    { isiMaxTime  :: !Timestamp -- ^ Effective 'IndexStateTime' state
+    , isiHeadTime :: !Timestamp -- ^ 'Timestamp' equivalent to 'IndexStateHead'
+    }
+
+emptyStateInfo :: IndexStateInfo
+emptyStateInfo = IndexStateInfo nullTimestamp nullTimestamp
+
+-- | 'IndexState' specification to 'Cache'
```

Comment:
Maybe better: Filters a 'Cache' according to an 'IndexState' specification. Also returns 'IndexStateInfo' describing the resulting index cache.